### DISCUSSION TITLE
Asset Automaterialization downstream of yielded AssetMaterializations

### DISCUSSION THREAD

**Discussion by matthias-Q** [dagster-io/dagster#18784] [Q&A] [answered]
Hi there. I have a problem with auto-materialization of a software defined asset that sits downstream of an asset that is materialized in a op yielding an AssetMaterialization. Here is some sample code:

```python
from dagster import (
    AssetKey,
    AssetMaterialization,
    AutoMaterializePolicy,
    DagsterUserCodeExecutionError,
    Output,
    asset,
    job,
    op,
)

@op
def op_yielding_asset(context):
    yield AssetMaterialization(
        asset_key=AssetKey(["debug", "upstream_asset"]),
    )

    yield Output(1, output_name="result")


@job
def asset_yielding_job():
    op_yielding_asset()


@asset(
    key_prefix=["debug"],
    deps={AssetKey(["debug", "upstream_asset"])},
    auto_materialize_policy=AutoMaterializePolicy.eager(),
)
def downstream_asset():
    return 1

```

The downstream asset only gets materialized once and never againg when the upstream asset is refreshed. Is this a bug or a configuration error on my side?

Thanks in advance,
Matt

**Comment by OwenKephart** [comment, marked as answer]
Hi @matthias-Q -- because there is no asset definition (i.e. `@asset`-decorated function) associated with `debug/upstream_asset`, Dagster assumes that it does not need to listen for AssetMaterialization events for that dependency.

My initial recommendation would be to just represent upstream asset as an `@asset`, but if that is not possible for some reason, the workaround would be to create a dummy `@asset` to represent `debug/upstream_asset`. This could have an empty body, but it would signify to the internals that it should expect asset materialization events to occur.

**Reply by matthias-Q** [reply]
Thanks for the response. This is a minimal example of one job that we have in production. It is a job that triggers an external process (legacy java code via SSH on a different server). In this process multiple tables are created (fan out of an op, each yielding an AssetMaterialization). So representing this with the `@asset` decorator is not possible.

This brings up the question for me, why the AssetMaterialization is there in the first place if it is not possible to have downstream depending assets. That feels a bit weird.

**Reply by OwenKephart** [reply]
The `AssetMaterialization` event pre-dates the `@asset` decorator -- originally, it just helped catalog events that occurred for a given asset, but were not associated with any specific code artifact. 

It sounds like the best option here is to just create a dummy asset and keep the job.

**Reply by matthias-Q** [reply]
So, when I define a dummy `upstream_asset` and then send the AssetMaterialization event from my op, dagster recognizes it and triggered the downstream asset?

**Reply by matthias-Q** [reply]
OK, I tested your workaround and it is a solution. But still, it feels a bit hacky. I will mark is as a solution.
